### PR TITLE
Create new load_functions_bbg rake task to only load cartodb extensio…

### PR DESCRIPTION
**Pull request by acarrera42**
…ns and quota updates on dedicated users and organization owners

During migration the following will need to be run:

bundle exec rake cartodb:db:load_functions_bbg
bundle exec rake cartodb:db:load_functions

### CODE REVIEW

File: lib/tasks/db_maintenance.rake [hunk 1]
```diff
@@ -234,8 +234,7 @@ namespace :cartodb do
         count = ::User.where(database_host: database_host).count
       end
       execute_on_users_with_index(task_name, Proc.new { |user, i|
-        begin
-          log(sprintf("Trying on %-#{20}s %-#{20}s (%-#{4}s/%-#{4}s)...", user.username, user.database_name, i+1, count), task_name, database_host)
+        begin log(sprintf("Trying on %-#{20}s %-#{20}s (%-#{4}s/%-#{4}s)...", user.username, user.database_name, i+1, count), task_name, database_host) # Only load cartodb extensions once per database, i.e. for dedicated users and org owners
```
Comment (tylerparsons) [inline lib/tasks/db_maintenance.rake]: Looks like this was just a formatting change, was this intentional?

File: lib/tasks/db_maintenance.rake [hunk 2]
```diff
@@ -251,6 +250,47 @@ namespace :cartodb do
       }, threads, thread_sleep, database_host)
     end
 
+    desc 'Install/upgrade CARTODB SQL functions - only organization owners for Bloomberg merge code release'
```
Comment (tylerparsons) [inline lib/tasks/db_maintenance.rake:254]: Let's not make this bloomberg specific, carto has mentioned they are also interested in this functionality. The real purpose of this is to only apply extensions once per database, something like `load_functions_for_all_dbs`.

File: lib/tasks/db_maintenance.rake [hunk 3]
```diff
@@ -251,6 +250,47 @@ namespace :cartodb do
...
+      end
+      execute_on_users_with_index(task_name, Proc.new { |user, i|
+        begin
+          log(sprintf("Trying on %-#{20}s %-#{20}s (%-#{4}s/%-#{4}s)...", user.username, user.database_name, i+1, count), task_name, database_host)
+          # Only load cartodb extensions once per database, i.e. for dedicated users and org owners
+          load_cartodb_extension = user.account_type == '[DEDICATED]' || user.organization_owner?
+          if load_cartodb_extension
+            log(sprintf("Loading cartodb extension and rebuild quota on %-#{20}s %-#{20}s (%-#{4}s/%-#{4}s)...", user.username, user.database_name, i+1, count), task_name, database_host)
+          end
+          user.db_service.load_cartodb_functions(statement_timeout, extension_version, load_cartodb_extension, load_cartodb_extension)
```
Comment (tylerparsons) [inline lib/tasks/db_maintenance.rake]: I don't think passing false for `rebuild_quota` from here is the right way to optimize this function.  This code will still spin off a thread and load user objects from the db for every single user. I think adding an additional filter to `execute_on_users_with_index` to only execute this block for users who are db owners is the way to go here.

**Review by tylerparsons — CHANGES_REQUESTED** [review]

File: lib/tasks/db_maintenance.rake (hunk 2 above)
Comment (acarrera42) [reply lib/tasks/db_maintenance.rake:254]: This function though is primarily to block the rebuilding of quotas.  I left the extension logic untouched in both rake tasks.  Is that also carto's interest?  Or are they only looking to take an upstream PR for the extension?

File: lib/tasks/db_maintenance.rake (hunk 3 above)
Comment (acarrera42) [reply lib/tasks/db_maintenance.rake]: That is a reasonable approach.  If this is something to be used upstream and more than the initial merge upgrade then it makes sense to modify this further.  If we put the user check outside of the loop then both rake tasks should be modified with this logic and the load_cartodb_extension and rebuild_quota parameters of load_cartodb_functions() should be removed.

File: lib/tasks/db_maintenance.rake (hunk 1 above)
Comment (acarrera42) [reply lib/tasks/db_maintenance.rake]: No, this should be reverted (though it doesn't affect functionality)

**Comment by vmanohar1** [comment]
If multi-threading worked correctly, we wouldn't need any change in code.. There is a bug in carto code here when using multi-threading..
See - https://github.com/CartoDB/cartodb/issues/13433

File: lib/tasks/db_maintenance.rake [hunk 4]
```diff
@@ -835,6 +872,61 @@ namespace :cartodb do
...
+      if database_host.nil?
+        puts "Detailed log stored at log/rake_db_maintenance_#{task_name}.log"
+      else
+        puts "Detailed log stored at log/rake_db_maintenance_#{task_name}_#{database_host}.log"
+      end
+
+      thread_pool = ThreadPool.new(num_threads, sleep_time)
+
+      if database_host.nil?
+        ::User.with_sql("select users.* from users, organizations where users.organization_id=organizations.id and (users.account_type = '[DEDICATED]' OR users.id = organizations.owner_id) order by created_at asc").each_with_index do |user, i|
```
Comment (tylerparsons) [inline lib/tasks/db_maintenance.rake]: Better to reuse this result set but fine for this use case

File: lib/tasks/db_maintenance.rake [hunk 5]
```diff
@@ -835,6 +872,61 @@ namespace :cartodb do
...
+    # Executes a ruby code proc/block on all existing users who are dedicated or organization owners, outputting some info
+    # @param task_name string
+    # @param block Proc
+    # @example:
+    # execute_on_dedicated_users_or_organization_owners_with_index(:populate_new_fields.to_s, Proc.new { |user, i| ... })
+    def execute_on_dedicated_users_or_organization_owners_with_index(task_name, block, num_threads=1, sleep_time=0.1, database_host=nil)
+      puts 'actest'
+      if database_host.nil?
+        puts "NO database host"
+        count = ::User.with_sql("select users.* from users, organizations where users.organization_id=organizations.id and (users.account_type = '[DEDICATED]' OR users.id = organizations.owner_id)").count
```
Comment (tylerparsons) [inline lib/tasks/db_maintenance.rake]: Some dedicated users may not have an organization. This should be a left join on organizations to pick up dedicated users with null `organization_id`.

**Review by tylerparsons — APPROVED** [review]

**Pull request merged**